### PR TITLE
FeeSub contract

### PR DESCRIPTION
**Pull request by bilbeyt**
Close: #2020

### CODE REVIEW

File: contracts/contracts/FeeSub.sol [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+            targetAddress,
+            amount,
+            validityPeriod
+        );
+        senders[requestId] = msg.sender;
+
+        return requestId;
+    }
+
+    function transferableAmount(
```
Comment (fredo) [inline contracts/contracts/FeeSub.sol:62]: We figured out that we don't need this function

File: contracts/contracts/FeeSub.sol [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+            ,
+            ,
+            ,
+
+        ) = requestManager.requests(requestId);
+        require(
+            sender == address(this),
+            "Request was not created by this contract"
+        );
+        require(senders[requestId] != address(0), "Already Withdrawn");
```
Comment (fredo) [inline contracts/contracts/FeeSub.sol:97]: nitpick `Already withdrawn`

File: contracts/contracts/FeeSub.sol [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+        IERC20 token = IERC20(sourceTokenAddress);
+        token.safeTransfer(recipient, amount);
+    }
+
+    function setMinimumAmount(
+        address tokenAddress,
+        uint256 amount
+    ) external onlyOwner {
+        if (amount > 0) {
```
Comment (fredo) [inline contracts/contracts/FeeSub.sol:112]: We can save some gas here and ask if minimumAmounts[tokenAddress] is already > 0, then we know that we already approved and don't need to approve again.

File: contracts/contracts/FeeSub.sol [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+            IERC20(tokenAddress).approve(
+                address(requestManager),
+                type(uint256).max
+            );
+        } else if (amount == 0) {
```
Comment (fredo) [inline contracts/contracts/FeeSub.sol:117]: same here if it's already set to 0 (which is kinda nonsense)

File: contracts/contracts/FeeSub.sol [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+            ,
+
+        ) = requestManager.requests(requestId);
+        require(
+            sender == address(this),
+            "Request was not created by this contract"
+        );
+        require(senders[requestId] != address(0), "Already Withdrawn");
+
+        requestManager.withdrawExpiredRequest(requestId);
```
Comment (fredo) [inline contracts/contracts/FeeSub.sol:99]: Here we need to add another condition. It can be the case that the request was already withdrawn by calling the RequestManager directly. 
If that's the case this call would revert and the user's money would be stuck inside this contract forever. 

I suggest that we do this check by checking the withdrawnClaimId of the request. 
We should only call `requestManager.withdrawExpiredRequest(requestId);` if withdrawnClaimId is 0.

Comment (fredo) [reply contracts/contracts/FeeSub.sol:99]: Speaking of which, we also need an additional require that the tokens have not been withdrawn by the agent. 
IOW, we should only ever proceed if we either can call `withdrawExpiredRequest` or we are sure that the money was already withdrawn, but to this contract


Suggestion: 

```sol
if(withdrawClaimId == 0) {
    // This will fail if the funds do not belong to the user (yet)
    requestManager.withdrawExpiredRequest(requestId);
}
else {
    // Make sure that funds were withdrawn by calling withdrawExpiredRequest()
    require(withdrawnClaimId == type(uint96).max);

}


```

File: contracts/contracts/FeeSub.sol (hunk 1 above)
Comment (GabrielBuragev) [reply contracts/contracts/FeeSub.sol:62]: I am not sure about this one. Can you remind me why we decided to not use this fn ?

File: contracts/contracts/FeeSub.sol [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+            ,
+            uint96 withdrawClaimId,
+            ,
+
+        ) = requestManager.requests(requestId);
+        require(
+            sender == address(this),
+            "Request was not created by this contract"
+        );
+        require(senders[requestId] != address(0), "Already withdrawn");
```
Comment (fredo) [inline contracts/contracts/FeeSub.sol:80]: I would rephrase it to "already refunded to the sender".
The rationale behind this is that we should have a distinction between withdrawing from the request manager and withdrawing from this contract.

Below there is `Already withdrawn by agent` -> I would this rename to `request is withdrawn by another address` (wouldnt mention the notion agent here as agent refers more to the software component)

File: contracts/contracts/FeeSub.sol (hunk 1 above)
Comment (bilbeyt) [reply contracts/contracts/FeeSub.sol]: @fredo

Comment (fredo) [reply contracts/contracts/FeeSub.sol]: already explained him in DM. 
We do not need this function cause it's trivial. If we were to use this contract (which is if transfer amount >= minimumAmount) then the transferable amount = transfer amount. 
If the condition is false we are not using the contract anyway.

File: beamer/tests/contracts/test_fee_sub.py [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+
+    assert token.balanceOf(sender) == mint_amount
+
+    request = fee_sub.createRequest(
+        chain_id, token.address, token.address, target, sent_amount, 1800, sender=sender
+    )
+
+    request_id = request.return_value
+
+    assert token.balanceOf(sender) == mint_amount - sent_amount  # no fee
```
Comment (fredo) [inline beamer/tests/contracts/test_fee_sub.py:50]: can you also assert that the `fee_sub_balance == old_balance - sent_amount - total_fee(sent_amount)`

File: contracts/contracts/FeeSub.sol [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+        );
+
+        if (withdrawClaimId == 0) {
+            // This will fail if the funds do not belong to the original sender (yet)
+            requestManager.withdrawExpiredRequest(requestId);
+        } else {
+            // Make sure that funds were withdrawn by calling withdrawExpiredRequest()
+            require(
+                withdrawClaimId == type(uint96).max,
+                "Request is withdrawn by another address"
```
Comment (fredo) [inline contracts/contracts/FeeSub.sol:92]: nitpick: `was withdrawn`

File: beamer/tests/contracts/test_fee_sub.py (hunk 7 above)
Comment (bilbeyt) [reply beamer/tests/contracts/test_fee_sub.py:50]: is this really true? old balance - total_fee is ok to me and for the sent_amount we are already getting it and sending it so no storing in there

Comment (fredo) [reply beamer/tests/contracts/test_fee_sub.py:53]: ah you are right

**Review by fredo — APPROVED** [review]
looks good 👍
